compiler/rustc_target/src/callconv/x86_win32.rs

```diff
@@ -0,0 +1,79 @@
+use rustc_abi::{Align, HasDataLayout, Reg, TyAbiInterface};
+
+use crate::callconv::FnAbi;
+use crate::spec::HasTargetSpec;
+
+pub(crate) fn compute_abi_info<'a, Ty, C>(
+    cx: &C,
+    fn_abi: &mut FnAbi<'a, Ty>,
+    opts: super::x86::X86Options,
+) where
+    Ty: TyAbiInterface<'a, C> + Copy,
+    C: HasDataLayout + HasTargetSpec,
+{
+    if !fn_abi.ret.is_ignore() {
+        if fn_abi.ret.layout.is_aggregate() && fn_abi.ret.layout.is_sized() {
+            // Returning a structure. Most often, this will use
+            // a hidden first argument. On some platforms, though,
+            // small structs are returned as integers.
+            //
+            // Some links:
+            // https://www.angelcode.com/dev/callconv/callconv.html
+            // Clang's ABI handling is in lib/CodeGen/TargetInfo.cpp
+            let t = cx.target_spec();
+            if t.abi_return_struct_as_int || opts.reg_struct_return {
+                // According to Clang, everyone but MSVC returns single-element
+                // float aggregates directly in a floating-point register.
```

Comment:
Okay so this seemed slightly sus, but it was added in https://github.com/rust-lang/rust/pull/44066 and looks right. Original comment was

> For struct FloatOne { double x; }, it appears that MSVC passes the parameter in RCX and returns in RAX, but mingw-gcc passes the parameter and return value both in XMM0. Mingw-clang does the same as MSVC, so I think this is a GCC bug.

reported to gcc in https://gcc.gnu.org/bugzilla/show_bug.cgi?id=82028 marked as dupe of https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85667.

Reply:
Dunno if you want to add another link, up to you.